src/lem/multiframe.rs

```diff
@@ -61,6 +63,121 @@ impl<'a, F: LurkField, C: Coprocessor<F>> MultiFrame<'a, F, C> {
+                let output = func
+                    .synthesize_frame(
+                        &mut cs.namespace(|| format!("frame {i}")),
+                        store,
+                        frame,
+                        g,
+                        bound_allocations,
+                        self.get_lang(),
+                    )
+                    .expect("failed to synthesize frame");
+                assert_eq!(input.len(), output.len());
+                Ok((i + 1, output))
+            })?;
+        Ok(output)
+    }
+
+    fn synthesize_frames_parallel<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &Store<F>,
+        input: Vec<AllocatedPtr<F>>,
+        frames: &[Frame<F>],
+        g: &GlobalAllocator<F>,
+    ) -> Result<Vec<AllocatedPtr<F>>, SynthesisError> {
+        assert!(!frames.is_empty());
+        assert!(cs.is_witness_generator());
+        assert!(CONFIG.parallelism.synthesis.is_parallel());
+        const MIN_CHUNK_SIZE: usize = 10;
+
+        let num_frames = frames.len();
+
+        let chunk_size = CONFIG
+            .parallelism
+            .synthesis
+            .chunk_size(num_frames, MIN_CHUNK_SIZE);
+
+        // We partition the frames into chunks, ideally one for each CPU. Each chunk will produce its
+        // corresponding partial witness in parallel, which are then collected into a vector.
+        let css = frames
+            .par_chunks(chunk_size)
+            .enumerate()
+            .map(|(chunk_idx, chunk)| {
+                let mut cs = WitnessCS::new();
+                // The first chunk will take as input the actual input of the circuit.
+                // Subsequent chunks would have to take the output of the previous chunk as input.
+                // But since we know the values of each chunk input and we are generating the
+                // witnesses separately and in parallel, we will allocate new variables for each
+                // chunk.
+                let chunk_input = if chunk_idx == 0 {
+                    input.clone()
+                } else {
+                    let first_chunk_frame = &frames[chunk_idx * chunk_size];
+                    let mut chunk_input = Vec::with_capacity(input.len());
+                    for (input_ptr_idx, input_ptr) in first_chunk_frame.input.iter().enumerate() {
+                        let z_ptr = store.hash_ptr(input_ptr).expect("Hash pointer failed");
+                        let alloc_ptr = AllocatedPtr::alloc(
+                            &mut cs.namespace(|| format!("var: {chunk_idx}.{input_ptr_idx}")),
+                            || Ok(z_ptr),
+                        )
+                        .expect("Allocation failed");
+                        chunk_input.push(alloc_ptr)
+                    }
+                    chunk_input
+                };
+
+                let chunk_output =
+                    self.synthesize_frames_sequential(&mut cs, store, chunk_input, chunk, g);
```

Comment:
It would be nice to have a guardrail in the same style that the sequential algorithm has, asserting that the output conforms to the values that we expect